src/SliceEnumerators.cs

```diff
@@ -0,0 +1,128 @@
+        /// </summary>
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        /// <summary>
+        /// please use the generic version if possible
+        /// </summary>
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// compiler is using this method in every occurence of foreach
+        /// where Slice IS accessed via the <see cref="IEnumerable{T}"/> interface
+        /// </summary>
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// A struct-based enumerator, to make fast enumerations possible.
+        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// Using structure avoids expensive heap allocations
+        /// </summary>
+        public struct Enumerator
+        {
+            Slice<T> m_slice;    // The slice being enumerated.
+            int      m_position; // The current position.
+
+            internal Enumerator(Slice<T> slice)
+            {
+                m_slice = slice;
+                m_position = -1;
+            }
+
+            public T Current
+            {
+                get { return m_slice[m_position]; }
+            }
+
+            public bool MoveNext()
+            {
+                return ++m_position < m_slice.Length;
+            }
+        }
+
+        /// <summary>
+        /// enumerator that implements <see cref="IEnumerator{T}"/> pattern (including <see cref="IDisposable"/>).
+        /// it is used by LINQ and foreach when Slice is accessed via <see cref="IEnumerable{T}"/>
+        /// it is reference type to avoid boxing when calling interface methods on stuctures
+        /// </summary>
+        private class EnumeratorObject : IEnumerator<T>
+        {
+            Slice<T> m_slice;    // The slice being enumerated.
+            int      m_position; // The current position.
+
+            public EnumeratorObject(Slice<T> slice)
+            {
+                m_slice = slice;
+                m_position = -1;
+            }
+
+            public T Current
+            {
+                get { return m_slice[m_position]; }
+            }
+
+            object IEnumerator.Current
+            {
+                get { return Current; }
+            }
+
+            public void Dispose()
+            {
+                // we don't have any managed resources here so we do not Dispose anything here
+                // we also don't need to set m_slice to null because when this instance becames garbage
+                // the reference stored in this field is also going to became GC-reachable (garbage)
+            }
+
+            public bool MoveNext()
+            {
+                return ++m_position < m_slice.Length;
+            }
```

Comment:
I think IEnumerator is speced to tolerate additional MoveNext calls after reaching the end without incrementing further (or it might wrap around).

Note that this is not a requirement for the struct enumeratror since that only needs to conform to "foreach" pattern, so that one is ok.